plugins/inventory/hrobot.py

```diff
@@ -0,0 +1,174 @@
+strict: False
+keyed_groups:
+  # Add e.g. dc hosts
+  - key: hrobot_dc
+    separator: ""
+# Use the IP address to connect to the host
+compose:
+  ansible_host: hrobot_server_ip
+"""
+
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+from ansible.module_utils.urls import open_url
+from ansible.utils.display import Display
+from ansible.errors import AnsibleError
+
+import json
+
+from ansible_collections.community.hrobot.plugins.module_utils.robot import (
+    BASE_URL
+)
+
+display = Display()
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'hrobot'
+
+    def verify_file(self, path):
+        ''' return true/false if this is possibly a valid file for this plugin to consume '''
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            # base class verifies that file exists and is readable by current user
+            if path.endswith(('hrobot.yaml', 'hrobot.yml')):
+                valid = True
+            else:
+                display.debug("hrobot inventory filename must end with 'hrobot.yml' or 'hrobot.yaml'")
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+        servers = {}
+        config = self._read_config_data(path)
+        self.load_cache_plugin()
+        cache_key = self.get_cache_key(path)
+
+        # cache may be True or False at this point to indicate if the inventory is being refreshed
+        # get the user's cache option too to see if we should save the cache if it is changing
+        user_cache_setting = self.get_option('cache')
+
+        # read if the user has caching enabled and the cache isn't being refreshed
+        attempt_to_read_cache = user_cache_setting and cache
+        # update if the user has caching enabled and the cache is being refreshed; update this value to True if the cache has expired below
+        cache_needs_update = user_cache_setting and not cache
+
+        # attempt to read the cache if inventory isn't being refreshed and the user has caching enabled
+        if attempt_to_read_cache:
+            try:
+                servers = self._cache[cache_key]
+            except KeyError:
+                # This occurs if the cache_key is not in the cache or if the cache_key expired, so the cache needs to be updated
+                cache_needs_update = True
+        elif not cache_needs_update:
+            servers = self.get_servers()
+
+        if cache_needs_update:
+            servers = self.get_servers()
+
+            # set the cache
+            self._cache[cache_key] = servers
+
+        self.populate(servers)
+
+    def populate(self, servers):
+        filters = self.get_option("filters")
+        strict = self.get_option('strict')
+        for server in servers:
+            s = server["server"]
+            server_name = s["server_name"]
```

Comment:
Hmm, is server name always set? I know it is for the servers I manage, but that might be because I explicitly set it :) The docs (https://robot.your-server.de/doc/webservice/en.html#get-server) are not very specific here.

How about changing this to
```suggestion
            server_name = s.get("server_name") or s["server_ip"]
```
so that in case the server_name is not present, `None`, or empty, it will be replaced by the server's main IP?

Reply:
I check other API, and when you modify server, you need to pass `server_name` as required https://robot.your-server.de/doc/webservice/en.html#post-server-server-ip as an example

Reply:
strange, but here seems like unique id is `server_ip` https://robot.your-server.de/doc/webservice/en.html#get-reset :)

Reply:
good idea to have failover to server_ip.

Reply:
I haven't tried, but I guess you can even use the same server name for different servers.

Reply:
if you have the same server name this api call won't be possible I think https://robot.your-server.de/doc/webservice/en.html#post-server-server-ip

Reply:
Why? That API uses the server IP as the main identifier, and allows to set the server name to some value.

Reply:
I just tried, the Robot web UI let me use the same name for two different servers.

Reply:
ok, got! thanks for trying out!

Reply:
> Regarding non-duplicate `server_name`s: one could also add a test in the plugin which makes it fail if two servers have the same `server_name` (at least when it is the value used for `inventory_hostname`).

what do you think about concatanation with `server_name` + `server_number`? 
and `server_name` will be set as `hrobot_server_name`?

Reply:
```
            server_name = s.get("server_name") or s["server_ip"]
            server_name = "{0}-{1}".format(server_name, s["server_number"])
```

Reply:
GitHub is hided this one, cc @felixfontein

Reply:
This is ok for a unique name (though just using `server_number` or `server_ip` is better for that purpose), but horrible for users. `inventory_hostname` should be something that can be controlled by the user, like taking it from `server_name` by default. Having an option which allows to chose something else is ok, but using something like that by default seems wrong to me.

Reply:
AWS uses instance id as inventory_hostname and I was thinking that we could do the same. Otherwise, inventory_hostname could be ```server_name = s.get("server_name") or s["server_ip"]```. But there's a chance for user to that server could be override. 
1) option to have check for existing key in map, if found print warning to user to set inventory_hostname in plugin config to something different, but the choice is not so big :) 
2) option have check for existing key in map, if found print warning to user to force him to go to cloud console and change server name

Reply:
I think AWS instance IDs are more visible than Hetzner's `server_number`, which would be the analogue to insance IDs.

We could add a config option `hostname` with choices `server_number`, `server_ip` and `server_name`, and then use that field. If there is a collision, report an error and tell user to select something else (or fix their server names so they are unique, in case `server_name` was chosen - similar to your 2.).

What do you think about that?

Reply:
For me duplication of server names looks like a corner case, which could be easily fixed on user side. Let's go with option 2, if it will be the case, I will add logic with additional option in plugin config file.